Replace espree with babel.

sainthkh opened the pull request:
## Description

Replaced `espree` with `babel`. Fixes https://github.com/WordPress/gutenberg/issues/19952

## How has this been tested?

Ran `npm run docs:build` and found no change.

## Screenshots <!-- if applicable -->

N/A

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [N/A] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [N/A] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [N/A] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [N/A] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

sainthkh committed:
Replace espree with babel.

sainthkh committed:
Fix test failure.

sainthkh committed:
Fix lint.

nosolosw commented:
These are the testing steps I'm following;

- In `master`
  - Modify the JSDoc comment of some exported entity and use the optional chaining syntax. Suggestion: modify `getCurrentUser` in `packages/core-data/src/selectors.js`.
  - Run `npm install && npm run docs:build`.
  - The expected result is that it should fail due to the unrecognized syntax.
  - Clean up any changes.
- Switch to this branch:
  - Modify the JSDoc comment of some exported entity and use the optional chaining syntax. Suggestion: modify `getCurrentUser` in `packages/core-data/src/selectors.js`.
  - Run `npm install && npm run docs:build`
  - Expected result is that the appropiate docs will be updated. Following the suggestion above: there should be changes in `docs/designers-developers/developers/data/data-core.md` and `packages/core-data/README.md`.

As per these steps, this works as expected! Happy to see that the code changes are minimal. Going to review the code now.

nosolosw commented on packages/docgen/src/engine.js:
```diff
@@ -1,25 +1,27 @@
  * Internal dependencies.
  */
 const getIntermediateRepresentation = require( './get-intermediate-representation' );
+const babelConfig = require( '../../../babel.config' );
```
We shouldn't use files from outside the `packages/docgen` repository, otherwise, it'll fail when used as a standalone npm binary.

sainthkh replied:
Then, we can make a command line option like --babel-config. And pass the path to the file in `docs` folder.

How about this?

nosolosw commented on packages/docgen/src/engine.js:
```diff
@@ -1,25 +1,27 @@
 	} );
 
+	return babel.parse( source || '', {
+		presets,
```
It doesn't look like babel-parser admits a [preset option](https://babeljs.io/docs/en/babel-parser/#options)?

sainthkh replied:
Right, it should be removed.

sainthkh replied:
I remember it didn't work properly without this. I'll check it out by reading @babel/parser code.

nosolosw commented on packages/docgen/src/engine.js:
```diff
@@ -1,25 +1,27 @@
+		presets,
+		plugins: [ ...plugins, 'jsx' ],
+		sourceType: 'module',
+	} ).program; // unlike espree, the root of babel AST is File, not Program.
```
One thing that I usually do in these cases (comments that are useful during the review phase but perhaps don't make a lot of sense in the source code itself) is to add them myself as comments in the PR. Perhaps we could remove it?

sainthkh replied:
I'll do it.

oandregal reviewed (changes requested):
Good progress here, this is a nice well-scoped refactor!

We'd need to figure out a way for us to:

1) use the same configuration than Gutenberg without hard-coding the path
2) use the presets already available (that include JSX parsing)

sainthkh committed:
Remove config.

sainthkh commented:
After your comment about `presets` option, it made me wonder why it doesn't exist while `plugins` does. A babel `preset` is just a package of multiple `plugin`s. It's a bit weird to learn that you can set individual plugins while cannot set packages. 

After thinking about this weirdness for a while, I realized that we don't need `babel.config.js` at all.

When using babel, we expect it to transpile new syntax to old syntax. To do so, babel parses js code and runs plugins to generate new one. 

In other words, babel works in these 3 steps.

1. babel/parser parses code and generates AST.
2. plugins manipulates the AST and create new one.
3. babel generates new code from new AST.

As the jobs of babel/parser and babel plugins are different, we don't need to import `babel.config.js` at all. 

I removed other configs and removed them.

sainthkh committed:
Update package-lock.json

nosolosw commented on packages/docgen/src/engine.js:
```diff
@@ -1,24 +1,20 @@
-		},
+const getAST = ( source ) => {
+	return babel.parse( source || '', {
+		plugins: [ 'jsx' ],
```
For testing, I'm using this code that includes a variety of new syntax (nullish coalescencing, optional chaining, optional catch binding, and numerical separator):

```js
	try {
		return state?.currentUser;
	} catch {
		return state ?? 1_000;
	}
```

try adding it to the `getCurrentUser` selector within the `packages/core-data/src/selectors.js` file and see how `npm run docs:build` fails due to unrecognizing the numerical separator syntax. The same code works fine with `npm run build`.

To be fully compatible with our current setup, we need to add the [numerical separator proposal](https://babeljs.io/docs/en/next/babel-parser#ecmascript-proposals-https-githubcom-babel-proposals) as a plugin as well:

```suggestion
		plugins: [ 'jsx', 'numericalSeparator' ],
```

The reason is that our `@wordpress/babel-preset-default` config [makes use](https://github.com/WordPress/gutenberg/blob/master/packages/babel-preset-default/index.js#L20) of the [shippedProposals](https://babeljs.io/docs/en/babel-preset-env#shippedproposals) option, which, at the time of writing, also enables the numeric separator syntax ([see](https://github.com/babel/babel/blob/v7.9.5/packages/babel-compat-data/scripts/data/plugin-features.js#L124)).

nosolosw replied:
How about we add a comment here that says something along these lines: 

When upgrading babel, we need to make sure this list of plugins are updated to include the shippedProposals that we enable by default in `@wordpress/babel-preset-default` https://github.com/WordPress/gutenberg/blob/master/packages/babel-preset-default/index.js#L20

nosolosw commented:
Good point about the transforms plugins. We do need the syntax/parser plugins for the parser to understand some syntax potentially used in Gutenberg: in our case, the `jsx` plugin as well as the `numericalSeparator` proposal (see [full discussion](https://github.com/WordPress/gutenberg/pull/21853#discussion_r415847682)).

With that change, I think we're good to go.

nosolosw commented:
It does bother me a bit that we need to keep an additional babel config beyond the existing ones (wordpress/babel-preset-default + gutenberg babel.config.js). What if instead of using `babel-parser` we use `@babel/core` as we [do in Gutenberg](https://github.com/WordPress/gutenberg/blob/master/bin/packages/build-worker.js#L7). :thinking:

It looks like, that way, we could reuse `@wordpress/babel-preset-default` and won't need any extra config. If, at some point in the future, we needed also something was added to the Gutenberg's `babel.config.js` we could copy it verbatim (or even make it reusable), which sounds better than having to look into babel internals to see which proposals to enable, etc.

sainthkh commented:
I tested your proposal about `@babel/core` by adding a file at `docgen/src/t.js` with the code below:

```js
const core = require( '@babel/core' );
const parser = require( '@babel/parser' );

const code = `
try {
	state?.currentUser;
} catch {
	state ?? 1_000;
}
`;

try {
	core.parse( code );
	process.stdout.write( 'core passed\n' );
} catch ( e ) {
	process.stdout.write( `core failed: ${ e }\n` );
}

try {
	parser.parse( code, {
		plugins: [ 'jsx' ],
		sourceType: 'module',
	} );
	process.stdout.write( 'parser passed\n' );
} catch ( e ) {
	process.stdout.write( `parser failed: ${ e }\n` );
}

try {
	parser.parse( code, {
		plugins: [ 'jsx', 'numericSeparator' ],
		sourceType: 'module',
	} );
	process.stdout.write( 'parser2 passed\n' );
} catch ( e ) {
	process.stdout.write( `parser2 failed: ${ e }\n` );
}
```

Confirmed that `@babel/core` works correctly without any setting. Let's continue in that way.

sainthkh committed:
Use @babel/core instead.

nosolosw commented:
This is nice! We also get the Gutenberg root configs for free :clap: 

I think we're ready to merge, the only thing left would be to document how this works in the README. Perhaps we can add a section called "Babel configuration" after the "CLI Options" that says that it takes the [project-wide config](https://babeljs.io/docs/en/next/config-files#project-wide-configuration), what happens if there is none, etc. What do you think?

Also, could you test that documents are generated properly with the pre-commit hook? For some reason, the pre-commit hook is not working for me and I want to make sure the pre-commit hook doesn't mess up the current working directory babel uses to find the config.

Testing instructions would be:

- Paste the testing code within the `getCurrentUser` selector at `packages/core-data/src/selectors.js`.
- Modify some part of its JSON comment.
- `git add packages/core-data/src/selectors.js` and commit.
- The expected result is that the commit should fail because there are two new files to add (`docs/designers-developers/developers/data/data-core.md` and `packages/core-data/README.md`).

sainthkh committed:
Update doc.

sainthkh commented:
* I confirmed that pre-commit hook generates docs properly and commit fails.
* Updated documentation

nosolosw commented:
Good work here, @sainthkh!

youknowriad commented:
Impressive work here :)

nb commented:
Just wanted to say “Thank you”, @sainthkh  – I was working on some package changes including  and every commit was failing, because docgen did not supporting newer ECMASript features. Not the best experience you can imagine…

Today, I rebased with `master` and it was all fixed :)
